src/util/codec/mysql/json/json_modify.rs

```diff
@@ -0,0 +1,164 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::Json;
+use super::super::Result;
+use super::path_expr::{PathLeg, PathExpression};
+use std::mem;
+
+/// `ModifyType` is for modify a JSON.
+#[derive(Clone, Debug, PartialEq)]
+pub enum ModifyType {
+    /// `Insert` is for inserting a new element into a JSON.
+    Insert,
+    /// `Replace` is for replacing a old element from a JSON.
+    Replace,
+    /// `Set` = `Insert` | `Replace`
+    Set,
+}
+
+impl Json {
+    // Modifies a Json object by insert, replace or set.
+    // All path expressions cannot contain * or ** wildcard.
+    // If any error occurs, the input won't be changed.
+    pub fn modify(&mut self,
+                  path_expr_list: &[PathExpression],
+                  mut values: Vec<Json>,
+                  mt: ModifyType)
+                  -> Result<()> {
+        if path_expr_list.len() != values.len() {
+            return Err(box_err!("Incorrect parameter count"));
+        }
+        for expr in path_expr_list {
+            if expr.contains_any_asterisk() {
+                return Err(box_err!("Invalid path expression"));
+            }
+        }
+        for (expr, value) in path_expr_list.iter().zip(values.drain(..)) {
+            self.set_json(&expr.legs, value, &mt);
+        }
+        Ok(())
+    }
+
+    // `set_json` is used in Json::modify().
+    fn set_json(&mut self, path_legs: &[PathLeg], value: Json, mt: &ModifyType) {
```

Comment:
ModifyType is a plain enum. 
Maybe it's better to make it `Copy` and use `mt: ModifyType` here.